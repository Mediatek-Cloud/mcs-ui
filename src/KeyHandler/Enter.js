// @flow
import * as React from 'react';
import PropTypes from 'prop-types';
import KeyHandler from './KeyHandler';

export type Props = {
  onClick: (e: any) => void | Promise<void>,
};

const EVENT_NAME = 'keypress';
const KEY_NAME = 'Enter';

const Enter = ({ onClick }: Props) => (
  <KeyHandler eventName={EVENT_NAME} keyName={KEY_NAME} onClick={onClick} />
);

Enter.propTypes = {
  onClick: PropTypes.func.isRequired,
};

export default Enter;
