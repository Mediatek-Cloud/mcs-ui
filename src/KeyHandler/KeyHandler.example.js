// @flow
import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { KeyHandler, Enter, Esc } from '.';

storiesOf('KeyHandler', module)
  .add(
    'API',
    () => (
      <KeyHandler
        eventName="keyup"
        keyName="a"
        onClick={action('KeyHandler onClick')}
      />
    ),
    {
      info: {
        text: 'keyup a',
        inline: true,
        source: false,
      },
    },
  )
  .add('Enter', () => <Enter onClick={action('Enter onClick')} />, {
    info: {
      text: 'Enter',
      inline: true,
      source: false,
    },
  })
  .add('Esc', () => <Esc onClick={action('Esc onClick')} />, {
    info: {
      text: 'Esc',
      inline: true,
      source: false,
    },
  });
