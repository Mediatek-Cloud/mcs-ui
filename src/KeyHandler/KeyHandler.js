/* global window */
// @flow
import * as React from 'react';
import PropTypes from 'prop-types';

export type Props = {
  onClick: (e: any) => void | Promise<void>,
  eventName: string,
  keyName: string,
};

class KeyHandler extends React.PureComponent<Props> {
  static propTypes = {
    eventName: PropTypes.string.isRequired,
    keyName: PropTypes.string.isRequired,
  };

  componentDidMount() {
    const { handleKey } = this;
    const { eventName } = this.props;
    window.document.addEventListener(eventName, handleKey);
  }

  componentWillUnmount() {
    const { handleKey } = this;
    const { eventName } = this.props;
    window.document.removeEventListener(eventName, handleKey);
  }

  handleKey = ({ key }: KeyboardEvent) => {
    const { onClick, keyName } = this.props;
    if (key === keyName) {
      onClick();
    }
  };

  render() {
    return null;
  }
}

export default KeyHandler;
