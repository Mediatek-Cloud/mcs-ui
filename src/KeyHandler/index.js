// @flow
export { default } from './KeyHandler';
export { default as KeyHandler } from './KeyHandler';
export { default as Enter } from './Enter';
export { default as Esc } from './Esc';
