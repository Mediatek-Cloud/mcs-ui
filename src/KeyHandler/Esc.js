// @flow
import * as React from 'react';
import PropTypes from 'prop-types';
import KeyHandler from './KeyHandler';

export type Props = {
  onClick: (e: any) => void | Promise<void>,
};

const EVENT_NAME = 'keyup';
const KEY_NAME = 'Escape';

const Esc = ({ onClick }: Props) => (
  <KeyHandler eventName={EVENT_NAME} keyName={KEY_NAME} onClick={onClick} />
);

Esc.propTypes = {
  onClick: PropTypes.func.isRequired,
};

export default Esc;
