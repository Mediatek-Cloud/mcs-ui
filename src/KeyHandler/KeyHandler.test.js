/* global document */
// @flow

import React from 'react';
import { mount } from 'enzyme';
import KeyHandler from '.';

it('should add a "keyup" EventListener of document', () => {
  // ref: https://github.com/airbnb/enzyme/issues/426
  const documentEvents = {};
  (document: any).addEventListener = jest.fn((event, cb) => {
    documentEvents[event] = cb;
  });
  (document: any).removeEventListener = jest.fn();

  // Before mount
  expect(documentEvents).toEqual({});
  const wrapper = mount(
    <KeyHandler eventName="keyup" keyName="a" onClick={() => {}} />,
  );

  // After mount
  expect(documentEvents).toMatchSnapshot();

  // Before Unmount
  expect(document.removeEventListener).not.toHaveBeenCalled();
  // After Unmount
  wrapper.unmount();
  expect(document.removeEventListener).toHaveBeenCalled();
});

it('should handle onClick', () => {
  const documentEvents = {};
  (document: any).addEventListener = jest.fn((event, cb) => {
    documentEvents[event] = cb;
  });
  const mockOnClick = jest.fn();
  mount(<KeyHandler eventName="keyup" keyName="a" onClick={mockOnClick} />);

  // Before clicking
  expect(mockOnClick).not.toHaveBeenCalled();
  // After clicking
  documentEvents.keyup({ key: 'a' });
  expect(mockOnClick).toHaveBeenCalled();
});
