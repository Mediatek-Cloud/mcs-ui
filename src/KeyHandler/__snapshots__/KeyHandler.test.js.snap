// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add a "keyup" EventListener of document 1`] = `
Object {
  "keyup": [Function],
}
`;
